include/tscore/ArgParser.h

```diff
@@ -0,0 +1,252 @@
+/** @file
+
+  Powerful and easy-to-use command line parsing for ATS
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include <iostream>
+#include <string>
+#include <map>
+#include <vector>
+#include <functional>
+#include <string_view>
+
+// more than zero arguments
+static constexpr unsigned MORE_THAN_ZERO_ARG_N = ~0;
+// more than one arguments
+static constexpr unsigned MORE_THAN_ONE_ARG_N = ~0 - 1;
+// customizable indent for help message
+static constexpr int INDENT_ONE   = 24;
+static constexpr int INDENT_TWO   = 32;
+static constexpr int INDENT_THREE = 46;
+
+namespace ts
+{
+// Some type def
+using StringArray = std::vector<std::string>;
+using Function    = std::function<void()>;
```

Comment:
The potential problem is that another ts header might define these same types.  An alternative is to define a struct just for types and use private inheritance:

struct ArgumentTypeDefs
{
  typedef std::vector<std::string> StringArray;
  typedef std::function<void()> Function;
};

class ArgumentData : private ArgumentTypeDefs
...

Reply:
Good point

Reply:
Maybe I'll just define them inside the classes.